src/persist-client/src/internal/machine.rs

```diff
@@ -1172,4 +1175,50 @@ mod tests {
             max_live_diffs
         );
     }
+
+    // A regression test for #14719, where a bug in gc led to an incremental
+    // state invariant being violated which resulted in gc being permanently
+    // wedged for the shard.
+    #[tokio::test(flavor = "multi_thread")]
+    async fn regression_gc_skipped_req_and_interrupted() {
+        let mut client = new_test_client().await;
+        let intercept = InterceptHandle::default();
+        client.blob = Arc::new(InterceptBlob::new(
+            Arc::clone(&client.blob),
+            intercept.clone(),
+        )) as Arc<dyn Blob + Send + Sync>;
+        let (_, mut read) = client
+            .expect_open::<String, String, u64, i64>(ShardId::new())
+            .await;
+
+        // Create a new SeqNo
+        read.downgrade_since(&Antichain::from_elem(1)).await;
+        let new_seqno_since = read.machine.seqno();
+
+        // Start a GC in the background for some SeqNo range that is not
+        // contiguous compared to the last gc req (in this case, n/a) and then
+        // crash when it gets to the blob deletes. In the regression case, this
+        // renders the shard permanently un-gc-able.
+        assert!(new_seqno_since > SeqNo::minimum());
+        intercept.set_post_delete(Some(Arc::new(|_, _| panic!("boom"))));
+        let mut machine = read.machine.clone();
+        let gc = spawn(|| "", async move {
```

Comment:
why does this need to be inside `spawn`?

Reply:
it only happens to finish (no boom panic) in the good case because it doesn't happen to call blob delete, but relying on that seems brittle to me. added some commentary